### PR TITLE
Quote (escape) filenames.  Add .gitignore.

### PR DESCRIPTION
**Pull request by arthurgleckler**

### CODE REVIEW

**Comment by stsquad** [comment]
@arthurgleckler : thanks for your contribution

File: dired-rsync.el [hunk 1]
```diff
@@ -145,7 +145,8 @@ ssh/scp tramp connections."
                        (-flatten
                         (list dired-rsync-command
                               dired-rsync-options
-                              src-files dest)))))
+                              (mapcar #'shell-quote-argument src-files)
```
Comment (stsquad) [inline dired-rsync.el]: Have you actually got a failure case for this? I think the escaping is handled by the shell process and this has regressed my use case.

**Comment by stsquad** [comment]
So given the files:

    * -rw-r--r--  1 alex alex     17 Apr 25 16:28 file1.txt                                                                                                                                        
    * -rw-r--r--  1 alex alex     36 Apr 25 16:28 file 2.txt                                                                                                                                       
    * -rw-r--r--  1 alex alex     31 Apr 25 16:28 file+3.txt                                                                                                                                       

With your patch applied I get:

    rsync: link_stat "/home/alex/tmp/file\ 2.txt" failed: No such file or directory (2)                                                                                                            
    rsync: link_stat "/home/alex/tmp/file\+3.txt" failed: No such file or directory (2)                                                                                                            
    file1.txt                                                                                                                                                                                      
    ^M              0   0%    0.00kB/s    0:00:00  ^M             17 100%   16.60kB/s    0:00:00 (xfr#1, to-chk=0/1)                                                                               
                                                                                                                                                                                               
    sent 43 bytes  received 297 bytes  226.67 bytes/sec                                                                                                                                            
    total size is 17  speedup is 0.05                                                                                                                                                              
    rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1659)[generator=3.1.3]

**Comment by arthurgleckler** [comment]
That's strange.  With exactly the same filenames, it works for me with the patch.  I mark the files in Dired, then do:

  M-x dired-rsync RET /tmp/ RET

Without the patch, the file with a space gets an error.

I'm running GNU Emacs 26.0.50.1 on MacOS 10.13.4.

What versions and operating system are you running?

Can you tell where the shell escaping is happening for you?

**Comment by stsquad** [comment]
I'm running:
  GNU Emacs 25.3.1 (x86_64-pc-linux-gnu, X toolkit) of 2018-03-29
  Linux remote and local
  Bash Shells as default

**Comment by arthurgleckler** [comment]
I'm perplexed.  I've just tried the patched version on GNU/Linux (Ubuntu 16.04) with GNU Emacs 25.1.1, both with and without my initialization files.  It works there, too, in both cases, but fails without the patch.

Would you mind trying again with <emacs -Q>?  That will ensure that no customizations are loaded.

**Comment by stsquad** [comment]
Same behaviour when running with emacs -Q. Very odd indeed.

**Comment by arthurgleckler** [comment]
Could you step through the code using Edebug to see where the escaping and quoting are happening for you?

**Comment by stsquad** [comment]
@arthurgleckler I think I have understood the confusion now. Could you check the branch really-fix-escaping and let me know if it works for you?

**Comment by arthurgleckler** [comment]
Unfortunately, that breaks it for me.  Now there's no escaping for the local copying case.

**Comment by stsquad** [comment]
Ahh right - local to local.. hmm how best to re-arrange that code?

**Comment by stsquad** [comment]
@arthurgleckler it's ugly but I think what I just pushed works.

**Comment by arthurgleckler** [comment]
That looks great, and works for me in both the local and remote cases.  (I tested on Linux.)

I can't believe that I didn't notice the existing use of `shell-quote-argument' when I was discussing this change with you.  Sorry about that.

By the way, you might consider making `dired-rsync--is-remote-tramp-p' and `dired-rsync--quote-and-maybe-convert-from-tramp' use `with-parsed-tramp-file-name' rather than parsing the Tramp filename yourself.  That way, your code will track changes to Tramp syntax, etc.

**Comment by stsquad** [comment]
OK I'll merge this fix for now and I'll look at making it better once I have pushed the fix ;-)

**Pull request merged**